### ISSUE TITLE
nd4j-native:0.5.0:macosx-x86_64 is significantly slower than nd4j-native:0.4.0:macosx-x86_64

### ISSUE THREAD

**Issue by mbabadi** [deeplearning4j/nd4j#1218]
Total runtime with nd4j-native:0.5.0:macosx-x86_64 = 1m18.287s
Total runtime with nd4j-native:0.4.0:macosx-x86_64 = 0m35.134s

There is a ~ 2.2x slowdown. The CPU hot spots obtained from JProfiler reveals the culprit:

org.nd4j.nativeblas.NativeOps.execPairwiseTransformFloat(org.bytedeco.javacpp.PointerPointer, int, org.bytedeco.javacpp.Pointer, org.bytedeco.javacpp.Pointer, org.bytedeco.javacpp.Pointer, org.bytedeco.javacpp.Pointer, org.bytedeco.javacpp.Pointer, org.bytedeco.javacpp.Pointer, org.bytedeco.javacpp.Pointer)

Each calls takes ~ 200 microseconds and there are tons of them (I suppose every pointwise operation invokes this method). There is no method call as such with nd4j-native:0.4.0:macosx-x86_64.

**Comment by treo** [comment]
https://github.com/deeplearning4j/nd4j/issues/1179 Discusses the exactly same problem.

**Comment by raver119** [comment]
Thats very surprising, since macos never had simd support via openmp...

a) Can we have profiler screenshots?
b) Can we have some info about code you're using there, which could help us to reproduce issue?

**Comment by raver119** [comment]
@treo yea, but macos runs without simd. And that what makes issue strange.

**Comment by mbabadi** [comment]
a) @raver119 I will attach profiler screenshots in a moment (I need to redo them).
b) The code has lots of pointwise NDArray operations, mostly in place, some with duplication, and some matrix multiplications. Many of these pointwise operations are performed in parallel streams so it might be a case of overthreading (I don't know how JVM threads interact with JavaCPP threads).

**Comment by raver119** [comment]
Great. What the average size for your pairwise operands?

**Comment by mbabadi** [comment]
At the moment, mostly 50 x 1000 matrices; everything is done on a single machine. In production runs, it will increase to 300 x 2000 for each spark partition.

**Comment by mbabadi** [comment]
@raver119 here are profiler screenshots:

nd4j-native:0.4.0:macosx-x86_64 CPU hot spot profile:
![0 4 0](https://cloud.githubusercontent.com/assets/15305869/17676703/abc37d20-62fd-11e6-81be-65897d4c2af8.png)

nd4j-native:0.5.0:macosx-x86_64 CPU hot spot profile:
![0 5 0](https://cloud.githubusercontent.com/assets/15305869/17676707/aeeaa58c-62fd-11e6-936d-9cca29fce27a.png)

**Comment by mbabadi** [comment]
@raver119 here is some code examples:
https://github.com/broadinstitute/gatk-protected/blob/mb_prob_coverage_model/src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/TargetCoverageEMComputeBlockNDArray.java

For example, see:
public ImmutablePair<INDArray, INDArray> getBiasLatentPosteriorDataUnregularized(...)
private UnivariateFunction createPsiBrentObjectiveFunction(...)
private static final Function<Map<String, ? extends Duplicable>, ? extends Duplicable> calculate_WzzWT_st(...)

**Comment by mbabadi** [comment]
@raver119 I will also test it on linux-x86_64 and let you know. For the record, here's the relevant part of our gradle build setup:

def libnd4jOS = null
if (isMacOsX) {
    libnd4jOS = 'macosx-x86_64'
} else if (isWindows) {
    libnd4jOS = 'windows-x86_64'
} else if (isLinux) {
    libnd4jOS = 'linux-x86_64'
} else {
    throw new Exception('Unknown OS. ND4J will be unable to find platform-specific binaries and thus unable to run.')
}

dependencies {
    compile files("${System.properties['java.home']}/../lib/tools.jar")
    compile 'org.broadinstitute:gatk:4.alpha.1-199-g8377acb-20160713.142718-1'
    compile 'org.broadinstitute:hdf5-java-bindings:1.0.0-hdf5_2.11.0'
    compile 'com.github.wendykierp:JTransforms:3.1'
    compile 'org.nd4j:nd4j-native:0.5.0'
    compile 'org.nd4j:nd4j-native:0.5.0:macosx-x86_64'
    compile 'org.nd4j:nd4j-native:0.5.0:linux-x86_64'
    compile 'org.nd4j:nd4j-serde:0.5.0'

    testCompile 'org.testng:testng:6.8.8'
}

**Comment by raver119** [comment]
nd4j-native-platform isn't working for you there?

**Comment by mbabadi** [comment]
Huh, I hadn't updated this in a while -- just checked, it worked. By the way, now that we are at it, `compile 'org.nd4j:nd4j-serde:0.5.0'` doesn't work for me for an unknown reason (never worked since 0.3 actually). I had to manually include Nd4j kryo serializer/registrator to my code base. Any ideas about that?

**Comment by raver119** [comment]
create separate issue for that, let's not spread out too much.

50 x 1000 matrices. the same order, or different orders possible?

**Comment by mbabadi** [comment]
Here are the only sizes I have: (S, T)    (S, L)    (S, L, L)    (1, T)    (S, 1)
Here, S ~ 50, T ~ 1000, L ~ 10. Sometimes I transpose things, so (T, S) is also possible.

**Comment by saudet** [comment]
OpenMP does get used on Mac OS X when compiled with clang-omp. BTW, I made some changes related to that in this PR: https://github.com/deeplearning4j/libnd4j/pull/250  Maybe that's where the difference comes from. They increased the speed in my tests on Fedora 23, but I guess we should test this with more configurations :)

**Comment by raver119** [comment]
This issue should be fixed now.
